README.md
[![HUG](https://raw.github.com/hugapi/hug/develop/artwork/logo.png)](http://hug.rest)
===================

[![PyPI version](https://badge.fury.io/py/hug.svg)](http://badge.fury.io/py/hug)
[![Build Status](https://travis-ci.org/hugapi/hug.svg?branch=develop)](https://travis-ci.org/hugapi/hug)
[![Windows Build Status](https://ci.appveyor.com/api/projects/status/0h7ynsqrbaxs7hfm/branch/master?svg=true)](https://ci.appveyor.com/project/TimothyCrosley/hug)
[![Coverage Status](https://coveralls.io/repos/hugapi/hug/badge.svg?branch=develop&service=github)](https://coveralls.io/github/hugapi/hug?branch=master)
[![License](https://img.shields.io/github/license/mashape/apistatus.svg)](https://pypi.python.org/pypi/hug/)
[![Join the chat at https://gitter.im/timothycrosley/hug](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/timothycrosley/hug?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

_________________

[Read Latest Documentation](https://hugapi.github.io/hug/) - [Browse GitHub Code Repository](https://github.com/hugapi/hug)
_________________

hug aims to make developing Python driven APIs as simple as possible, but no simpler. As a result, it drastically simplifies Python API development.

hug's Design Objectives:

- Make developing a Python driven API as succinct as a written definition.
- The framework should encourage code that self-documents.
- It should be fast. A developer should never feel the need to look somewhere else for performance reasons.
- Writing tests for APIs written on-top of hug should be easy and intuitive.
- Magic done once, in an API framework, is better than pushing the problem set to the user of the API framework.
- Be the basis for next generation Python APIs, embracing the latest technology.

As a result of these goals, hug is Python 3+ only and built upon [Falcon's](https://github.com/falconry/falcon) high performance HTTP library

[![HUG Hello World Example](https://raw.github.com/hugapi/hug/develop/artwork/example.gif)](https://github.com/hugapi/hug/blob/develop/examples/hello_world.py)

Supporting hug development
===================
[Get professionally supported hug with the Tidelift Subscription](https://tidelift.com/subscription/pkg/pypi-hug?utm_source=pypi-hug&utm_medium=referral&utm_campaign=readme)

Professional support for hug is available as part of the [Tidelift
Subscription](https://tidelift.com/subscription/pkg/pypi-hug?utm_source=pypi-hug&utm_medium=referral&utm_campaign=readme).
Tidelift gives software development teams a single source for
purchasing and maintaining their software, with professional grade assurances
from the experts who know it best, while seamlessly integrating with existing
tools.

Installing hug
===================

Installing hug is as simple as:

```bash
pip3 install hug --upgrade
```

Ideally, within a [virtual environment](http://docs.python-guide.org/en/latest/dev/virtualenvs/).

Getting Started
===================

Build an example API with a simple endpoint in just a few lines.

```py
# filename: happy_birthday.py
"""A basic (single function) API written using hug"""
import hug


@hug.get('/happy_birthday')
def happy_birthday(name, age:hug.types.number=1):
    """Says happy birthday to a user"""
    return "Happy {age} Birthday {name}!".format(**locals())
```

To run, from the command line type:

```bash
hug -f happy_birthday.py
```

You can access the example in your browser at:
`localhost:8000/happy_birthday?name=hug&age=1`. Then check out the
documentation for your API at `localhost:8000/documentation`

Parameters can also be encoded in the URL (check
out [`happy_birthday.py`](examples/happy_birthday.py) for the whole
example).

```py
@hug.get('/greet/{event}')
def greet(event: str):
    """Greets appropriately (from http://blog.ketchum.com/how-to-write-10-common-holiday-greetings/)  """
    greetings = "Happy"
    if event == "Christmas":
        greetings = "Merry"
    if event == "Kwanzaa":
        greetings = "Joyous"
    if event == "wishes":
        greetings = "Warm"

    return "{greetings} {event}!".format(**locals())
```

Which, once you are running the server as above, you can use this way:

```
curl http://localhost:8000/greet/wishes
"Warm wishes!"
```

Versioning with hug
===================

```py
# filename: versioning_example.py
"""A simple example of a hug API call with versioning"""
import hug

@hug.get('/echo', versions=1)
def echo(text):
    return text


@hug.get('/echo', versions=range(2, 5))
def echo(text):
    return "Echo: {text}".format(**locals())
```

To run the example:

```bash
hug -f versioning_example.py
```

Then you can access the example from `localhost:8000/v1/echo?text=Hi` / `localhost:8000/v2/echo?text=Hi` Or access the documentation for your API from `localhost:8000`

Note: versioning in hug automatically supports both the version header as well as direct URL based specification.

Testing hug APIs
===================

hug's `http` method decorators don't modify your original functions. This makes testing hug APIs as simple as testing any other Python functions. Additionally, this means interacting with your API functions in other Python code is as straight forward as calling Python only API functions. hug makes it easy to test the full Python stack of your API by using the `hug.test` module:

```python
import hug
import happy_birthday

hug.test.get(happy_birthday, 'happy_birthday', {'name': 'Timothy', 'age': 25}) # Returns a Response object
```

You can use this `Response` object for test assertions (check
out [`test_happy_birthday.py`](examples/test_happy_birthday.py) ):

```python
def tests_happy_birthday():
    response = hug.test.get(happy_birthday, 'happy_birthday', {'name': 'Timothy', 'age': 25})
    assert response.status == HTTP_200
    assert response.data is not None
```

Running hug with other WSGI based servers
===================

hug exposes a `__hug_wsgi__` magic method on every API module automatically. Running your hug based API on any standard wsgi server should be as simple as pointing it to `module_name`: `__hug_wsgi__`.

For Example:

```bash
uwsgi --http 0.0.0.0:8000 --wsgi-file examples/hello_world.py --callable __hug_wsgi__
```

To run the hello world hug example API.

Building Blocks of a hug API
===================

When building an API using the hug framework you'll use the following concepts:

**METHOD Decorators** `get`, `post`, `update`, etc HTTP method decorators that expose your Python function as an API while keeping your Python method unchanged

```py
@hug.get() # <- Is the hug METHOD decorator
def hello_world():
    return "Hello"
```

hug uses the structure of the function you decorate to automatically generate documentation for users of your API. hug always passes a request, response, and api_version variable to your function if they are defined params in your function definition.

**Type Annotations** functions that optionally are attached to your methods arguments to specify how the argument is validated and converted into a Python type

```py
@hug.get()
def math(number_1:int, number_2:int): #The :int after both arguments is the Type Annotation
    return number_1 + number_2
```

Type annotations also feed into `hug`'s automatic documentation
generation to let users of your API know what data to supply.

**Directives** functions that get executed with the request / response data based on being requested as an argument in your api_function.
These apply as input parameters only, and can not be applied currently as output formats or transformations.

```py
@hug.get()
def test_time(hug_timer):
    return {'time_taken': float(hug_timer)}
```

Directives may be accessed via an argument with a `hug_` prefix, or by using Python 3 type annotations. The latter is the more modern approach, and is recommended. Directives declared in a module can be accessed by using their fully qualified name as the type annotation (ex: `module.directive_name`).

Aside from the obvious input transformation use case, directives can be used to pipe data into your API functions, even if they are not present in the request query string, POST body, etc. For an example of how to use directives in this way, see the authentication example in the examples folder.

Adding your own directives is straight forward:

```py
@hug.directive()
def square(value=1, **kwargs):
    '''Returns passed in parameter multiplied by itself'''
    return value * value

@hug.get()
@hug.local()
def tester(value: square=10):
    return value

tester() == 100
```

For completeness, here is an example of accessing the directive via the magic name approach:

```py
@hug.directive()
def multiply(value=1, **kwargs):
    '''Returns passed in parameter multiplied by itself'''
    return value * value

@hug.get()
@hug.local()
def tester(hug_multiply=10):
    return hug_multiply

tester() == 100
```

**Output Formatters** a function that takes the output of your API function and formats it for transport to the user of the API.

```py
@hug.default_output_format()
def my_output_formatter(data):
    return "STRING:{0}".format(data)

@hug.get(output=hug.output_format.json)
def hello():
    return {'hello': 'world'}
```

as shown, you can easily change the output format for both an entire API as well as an individual API call

**Input Formatters** a function that takes the body of data given from a user of your API and formats it for handling.

```py
@hug.default_input_format("application/json")
def my_input_formatter(data):
    return ('Results', hug.input_format.json(data))
```

Input formatters are mapped based on the `content_type` of the request data, and only perform basic parsing. More detailed parsing should be done by the Type Annotations present on your `api_function`

**Middleware** functions that get called for every request a hug API processes

```py
@hug.request_middleware()
def process_data(request, response):
    request.env['SERVER_NAME'] = 'changed'

@hug.response_middleware()
def process_data(request, response, resource):
    response.set_header('MyHeader', 'Value')
```

You can also easily add any Falcon style middleware using:

```py
__hug__.http.add_middleware(MiddlewareObject())
```

**Parameter mapping** can be used to override inferred parameter names, eg. for reserved keywords:

```py
import marshmallow.fields as fields
...

@hug.get('/foo', map_params={'from': 'from_date'})  # API call uses 'from'
def get_foo_by_date(from_date: fields.DateTime()):
    return find_foo(from_date)
```

Input formatters are mapped based on the `content_type` of the request data, and only perform basic parsing. More detailed parsing should be done by the Type Annotations present on your `api_function`

Splitting APIs over multiple files
===================

hug enables you to organize large projects in any manner you see fit. You can import any module that contains hug decorated functions (request handling, directives, type handlers, etc) and extend your base API with that module.

For example:

`something.py`

```py
import hug

@hug.get('/')
def say_hi():
    return 'hello from something'
```

Can be imported into the main API file:

`__init__.py`

```py
import hug
from . import something

@hug.get('/')
def say_hi():
    return "Hi from root"

@hug.extend_api('/something')
def something_api():
    return [something]
```

Or alternatively - for cases like this - where only one module is being included per a URL route:

```py
#alternatively
hug.API(__name__).extend(something, '/something')
```

Configuring hug 404
===================

By default, hug returns an auto generated API spec when a user tries to access an endpoint that isn't defined. If you would not like to return this spec you can turn off 404 documentation:

From the command line application:

```bash
hug -nd -f {file} #nd flag tells hug not to generate documentation on 404
```

Additionally, you can easily create a custom 404 handler using the `hug.not_found` decorator:

```py
@hug.not_found()
def not_found_handler():
    return "Not Found"
```

This decorator works in the same manner as the hug HTTP method decorators, and is even version aware:

```py
@hug.not_found(versions=1)
def not_found_handler():
    return ""

@hug.not_found(versions=2)
def not_found_handler():
    return "Not Found"
```

Asyncio support
===============

When using the `get` and `cli` method decorator on coroutines, hug will schedule
the execution of the coroutine.

Using asyncio coroutine decorator

```py
@hug.get()
@asyncio.coroutine
def hello_world():
    return "Hello"
```

Using Python 3.5 async keyword.

```py
@hug.get()
async def hello_world():
    return "Hello"
```

NOTE: Hug is running on top Falcon which is not an asynchronous server. Even if using
asyncio, requests will still be processed synchronously.

Using Docker
===================

If you like to develop in Docker and keep your system clean, you can do that but you'll need to first install [Docker Compose](https://docs.docker.com/compose/install/).

Once you've done that, you'll need to `cd` into the `docker` directory and run the web server (Gunicorn) specified in `./docker/gunicorn/Dockerfile`, after which you can preview the output of your API in the browser on your host machine.

```bash
$ cd ./docker
# This will run Gunicorn on port 8000 of the Docker container.
$ docker-compose up gunicorn

# From the host machine, find your Dockers IP address.
# For Windows & Mac:
$ docker-machine ip default

# For Linux:
$ ifconfig docker0 | grep 'inet' | cut -d: -f2 | awk '{ print $1}' | head -n1
```

By default, the IP is 172.17.0.1. Assuming that's the IP you see, as well, you would then go to `http://172.17.0.1:8000/` in your browser to view your API.

You can also log into a Docker container that you can consider your work space. This workspace has Python and Pip installed so you can use those tools within Docker. If you need to test the CLI interface, for example, you would use this.

```bash
$ docker-compose run workspace bash
```

On your Docker `workspace` container, the `./docker/templates` directory on your host computer is mounted to `/src` in the Docker container. This is specified under `services` > `app` of `./docker/docker-compose.yml`.

```bash
bash-4.3# cd /src
bash-4.3# tree
.
├── __init__.py
└── handlers
    ├── birthday.py
    └── hello.py

1 directory, 3 files
```

Security contact information
===================

hug takes security and quality seriously. This focus is why we depend only on thoroughly tested components and utilize static analysis tools (such as bandit and safety) to verify the security of our code base.
If you find or encounter any potential security issues, please let us know right away so we can resolve them.

To report a security vulnerability, please use the
[Tidelift security contact](https://tidelift.com/security).
Tidelift will coordinate the fix and disclosure.

Why hug?
===================

HUG simply stands for Hopefully Useful Guide. This represents the project's goal to help guide developers into creating well written and intuitive APIs.

--------------------------------------------

Thanks and I hope you find *this* hug helpful as you develop your next Python API!

~Timothy Crosley


CHANGELOG.md
Install the latest
===================

To install the latest version of hug simply run:

```bash
pip3 install hug --upgrade
```

Ideally, within a virtual environment.

Changelog
=========
### 2.6.1 - February 6, 2020
- Fixed issue #889: Added support of Python 3.8
- Fixed issue #834: Bug in some cases when introspecting local documentation.

### 2.6.0 - August 29, 2019
- Improved CLI multiple behaviour with empty defaults
- Improved CLI type output for built-in types
- Improved MultiCLI base documentation

### 2.5.6 - June 20, 2019
- Fixed issue #815: map_params() causes api documentation to lose param type information
- Improved project testing: restoring 100% coverage

### 2.5.5 - June 13, 2019
- Fixed issue #808: Problems with command line invocation via hug CLI
- Fixed issue #647: Support for arbitrary URL complexity when using CORS middleware
- Fixed issue #805: Added documentation for `map_params` feature
- Added initial automated code cleaning and linting partially satisfying [HOPE-8 -- Style Guideline for Hug](https://github.com/hugapi/HOPE/blob/master/all/HOPE-8--Style-Guide-for-Hug-Code.md#hope-8----style-guide-for-hug-code)
- Implemented [HOPE-20 -- The Zen of Hug](https://github.com/hugapi/HOPE/blob/master/all/HOPE-20--The-Zen-of-Hug.md)

### 2.5.4 hotfix - May 19, 2019
- Fix issue #798 - Development runner `TypeError` when executing cli

### 2.5.3 hotfix - May 15, 2019
- Fixed issue #794 - Restore support for versions of Marshmallow pre-2.17.0

### 2.5.2 hotfix - May 10, 2019
- Fixed issue #790 - Set Falcon defaults to pre 2.0.0 version to avoid breaking changes for Hug users until a Hug 3.0.0 release. The new default Falcon behaviour can be defaulted before hand by setting `__hug__.future = True`.

### 2.5.1 hotfix - May 9, 2019
- Fixed issue #784 - POST requests broken on 2.5.0
- Optimizations and simplification of async support, taking advantadge of Python3.4 deprecation.
- Fix issue #785: Empty query params are not ignored on 2.5.0
- Added support for modifying falcon API directly on startup
- Initial `black` formatting of code base, in preperation for CI enforced code formatting

### 2.5.0 - May 4, 2019
- Updated to latest Falcon: 2.0.0
- Added support for Marshmallow 3
- Added support for `args` decorator parameter to optionally specify type transformations separate from annotations
- Added support for tests to provide a custom host parameter
- Breaking Changes:
    - Deprecated support for Python 3.4

### 2.4.9 - TBD
- Add the ability to invoke the hug development server as a Python module e.g. `python -m hug`
- Corrected the documentation for the `--without-cython` install option

### 2.4.8 - April 7, 2019
- Fixed issue #762 - HTTP errors crash with selectable output types
- Fixed MacOS testing via travis - added testing accross all the same Python versions tested on Linux

### 2.4.7 - March 28, 2019
- Fixed API documentation with selectable output types

### 2.4.6 - March 25, 2019
- Fixed issue #753 - 404 not found does not respect default output format.
- Documented the `--without-cython` option in `CONTRIBUTING.md`
- Extended documentation for output formats

### 2.4.4 - March 21, 2019
- Added the ability to change the default output format for CLI endpoints both at the API and global level.
- Added the ablity to extend CLI APIs in addition to HTTP APIs issue #744.
- Added optional built-in API aware testing for CLI commands.
- Add unit test for `extend_api()` with CLI commands
- Fix running tests using `python setup.py test`
- Fix issue #749 extending API with mixed GET/POST methods
- Documented the `multiple_files` example
- Added the `--without-cython` option to `setup.py`

### 2.4.3 [hotfix] - March 17, 2019
- Fix issue #737 - latest hug release breaks meinheld worker setup

### 2.4.2 - March 16, 2019
- Python 3.7 support improvements
- No longer test against Python 3.4 - aimed for full deprecation in Hug 3.0.0
- Improved interoperability with the latest Falcon release
- Documentation improvements
- Fixed bug in auto reload

### 2.4.1 - Sep 17, 2018
- Fixed issue #631: Added support for Python 3.7
- Fixed issue #665: Fixed problem with hug.types.json
- Fixed issue #679: Return docs for marshmallow schema instead of for dump method

### 2.4.0 - Jan 31, 2018
- Updated Falcon requirement to 1.4.1
- Fixed issue #590: Textual output formats should have explicitly defined charsets by default
- Fixed issue #596: Host argument for development runner
- Fixed issue #563: Added middleware to handle CORS requests
- Implemented issue #612: Add support for numpy types in JSON output by default
- Implemented improved class based directives with cleanup support (see: https://github.com/timothycrosley/hug/pull/603)
- Support ujson if installed
- Implement issue #579: Allow silencing intro message when running hug from command line
- Implemented issue #531: Allow exit value to alter status code for CLI tools
- Updated documentation generation to use hug's JSON outputter for consistency

### 2.3.2 - Sep 28, 2017
- Implemented Issue #540: Add support for remapping parameters
- Updated Falcon requirement to 1.3.0
- Fixed issue #552: Version ignored in class based routes
- Fixed issue #555: Gracefully handle digit only version strings
- Fixed issue #519: Exceptions are now correctly inserted into the current API using `extend_api`
- Breaking Changes:
    - Fixed issue #539: Allow JSON output to include non-ascii (UTF8) characters by default.

### 2.3.1 - Aug 26, 2017
- Fixed issue #500 & 504: Added support for automatic reload on Windows & enabled intuitive use of pdb within autoreloader
- Implemented improved way to retrieve list of urls and handlers for issue #462
- Implemented support for Python typing module style sub types
- Updated to allow -m parameter load modules on current directory
- Improved hug.test decode behaviour
- Added built in handlers for CORS support:
    - directive `hug.directives.cors`
    - Improved routing support
    - Added allow origins middleware


hug/use.py
"""hug/use.py

Provides a mechanism for using external hug APIs both locally or remotely in a seamless fashion

Copyright (C) 2016  Timothy Edmund Crosley

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
documentation files (the "Software"), to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or
substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

"""
from __future__ import absolute_import

import re
import socket
from collections import namedtuple
from io import BytesIO
from queue import Queue

import falcon
import requests

import hug._empty as empty
from hug.api import API
from hug.defaults import input_format
from hug.format import parse_content_type

Response = namedtuple("Response", ("data", "status_code", "headers"))
Request = namedtuple("Request", ("content_length", "stream", "params"))


class Service(object):
    """Defines the base concept of a consumed service.
        This is to enable encapsulating the logic of calling a service so usage can be independant of the interface
    """

    __slots__ = ("timeout", "raise_on", "version")

    def __init__(self, version=None, timeout=None, raise_on=(500,), **kwargs):
        self.version = version
        self.timeout = timeout
        self.raise_on = raise_on if type(raise_on) in (tuple, list) else (raise_on,)

    def request(
        self, method, url, url_params=empty.dict, headers=empty.dict, timeout=None, **params
    ):
        """Calls the service at the specified URL using the "CALL" method"""
        raise NotImplementedError("Concrete services must define the request method")

    def get(self, url, url_params=empty.dict, headers=empty.dict, timeout=None, **params):
        """Calls the service at the specified URL using the "GET" method"""
        return self.request("GET", url=url, headers=headers, timeout=timeout, **params)

    def post(self, url, url_params=empty.dict, headers=empty.dict, timeout=None, **params):
        """Calls the service at the specified URL using the "POST" method"""
        return self.request("POST", url=url, headers=headers, timeout=timeout, **params)

    def delete(self, url, url_params=empty.dict, headers=empty.dict, timeout=None, **params):
        """Calls the service at the specified URL using the "DELETE" method"""
        return self.request("DELETE", url=url, headers=headers, timeout=timeout, **params)

    def put(self, url, url_params=empty.dict, headers=empty.dict, timeout=None, **params):
        """Calls the service at the specified URL using the "PUT" method"""
        return self.request("PUT", url=url, headers=headers, timeout=timeout, **params)

    def trace(self, url, url_params=empty.dict, headers=empty.dict, timeout=None, **params):
        """Calls the service at the specified URL using the "TRACE" method"""
        return self.request("TRACE", url=url, headers=headers, timeout=timeout, **params)

    def patch(self, url, url_params=empty.dict, headers=empty.dict, timeout=None, **params):
        """Calls the service at the specified URL using the "PATCH" method"""
        return self.request("PATCH", url=url, headers=headers, timeout=timeout, **params)

    def options(self, url, url_params=empty.dict, headers=empty.dict, timeout=None, **params):
        """Calls the service at the specified URL using the "OPTIONS" method"""
        return self.request("OPTIONS", url=url, headers=headers, timeout=timeout, **params)

    def head(self, url, url_params=empty.dict, headers=empty.dict, timeout=None, **params):
        """Calls the service at the specified URL using the "HEAD" method"""
        return self.request("HEAD", url=url, headers=headers, timeout=timeout, **params)

    def connect(self, url, url_params=empty.dict, headers=empty.dict, timeout=None, **params):
        """Calls the service at the specified URL using the "CONNECT" method"""
        return self.request("CONNECT", url=url, headers=headers, timeout=timeout, **params)


class HTTP(Service):
    __slots__ = ("endpoint", "session", "json_transport")

    def __init__(
        self,
        endpoint,
        auth=None,
        version=None,
        headers=empty.dict,
        timeout=None,
        raise_on=(500,),
        json_transport=True,
        **kwargs
    ):
        super().__init__(timeout=timeout, raise_on=raise_on, version=version, **kwargs)
        self.endpoint = endpoint
        self.session = requests.Session()
        self.session.auth = auth
        self.session.headers.update(headers)
        self.json_transport = json_transport

    def request(
        self, method, url, url_params=empty.dict, headers=empty.dict, timeout=None, **params
    ):
        url = "{0}/{1}".format(self.version, url.lstrip("/")) if self.version else url
        kwargs = {"json" if self.json_transport else "params": params}
        response = self.session.request(
            method, self.endpoint + url.format(url_params), headers=headers, **kwargs
        )

        data = BytesIO(response.content)
        content_type, content_params = parse_content_type(response.headers.get("content-type", ""))
        if content_type in input_format:
            data = input_format[content_type](data, **content_params)

        if response.status_code in self.raise_on:
            raise requests.HTTPError(
                "{0} {1} occured for url: {2}".format(response.status_code, response.reason, url)
            )

        return Response(data, response.status_code, response.headers)


class Local(Service):
    __slots__ = ("api", "headers")

    def __init__(
        self, api, version=None, headers=empty.dict, timeout=None, raise_on=(500,), **kwargs
    ):
        super().__init__(timeout=timeout, raise_on=raise_on, version=version, **kwargs)
        self.api = API(api)
        self.headers = headers

    def request(
        self, method, url, url_params=empty.dict, headers=empty.dict, timeout=None, **params
    ):
        function = self.api.http.versioned.get(self.version, {}).get(url, None)
        if not function:
            function = self.api.http.versioned.get(None, {}).get(url, None)

        if not function:
            if 404 in self.raise_on:
                raise requests.HTTPError("404 Not Found occured for url: {0}".format(url))
            return Response("Not Found", 404, {"content-type", "application/json"})

        interface = function.interface.http
        response = falcon.Response()
        request = Request(None, None, empty.dict)
        context = self.api.context_factory(
            api=self.api, api_version=self.version, interface=interface
        )
        interface.set_response_defaults(response)

        params.update(url_params)
        params = interface.gather_parameters(
            request, response, context, api_version=self.version, **params
        )
        errors = interface.validate(params, context)
        if errors:
            interface.render_errors(errors, request, response)
        else:
            interface.render_content(interface.call_function(params), context, request, response)

        data = BytesIO(response.data)
        content_type, content_params = parse_content_type(response._headers.get("content-type", ""))
        if content_type in input_format:
            data = input_format[content_type](data, **content_params)

        status_code = int("".join(re.findall(r"\d+", response.status)))
        if status_code in self.raise_on:
            raise requests.HTTPError("{0} occured for url: {1}".format(response.status, url))

        return Response(data, status_code, response._headers)


class Socket(Service):
    __slots__ = ("connection_pool", "timeout", "connection", "send_and_receive")

    on_unix = getattr(socket, "AF_UNIX", False)
    Connection = namedtuple("Connection", ("connect_to", "proto", "sockopts"))
    protocols = {
        "tcp": (socket.AF_INET, socket.SOCK_STREAM),
        "udp": (socket.AF_INET, socket.SOCK_DGRAM),
    }
    streams = set(("tcp",))
    datagrams = set(("udp",))
    inet = set(("tcp", "udp"))
    unix = set()

    if on_unix:
        protocols.update(
            {
                "unix_dgram": (socket.AF_UNIX, socket.SOCK_DGRAM),
                "unix_stream": (socket.AF_UNIX, socket.SOCK_STREAM),
            }
        )
        streams.add("unix_stream")
        datagrams.add("unix_dgram")
        unix.update(("unix_stream", "unix_dgram"))

    def __init__(
        self,
        connect_to,
        proto,
        version=None,
        headers=empty.dict,
        timeout=None,
        pool=0,
        raise_on=(500,),
        **kwargs
    ):
        super().__init__(timeout=timeout, raise_on=raise_on, version=version, **kwargs)
        connect_to = tuple(connect_to) if proto in Socket.inet else connect_to
        self.timeout = timeout
        self.connection = Socket.Connection(connect_to, proto, set())
        self.connection_pool = Queue(maxsize=pool if pool else 1)

        if proto in Socket.streams:
            self.send_and_receive = self._stream_send_and_receive
        else:
            self.send_and_receive = self._dgram_send_and_receive

    def settimeout(self, timeout):
        """Set the default timeout"""
        self.timeout = timeout

    def setsockopt(self, *sockopts):
        """Add socket options to set"""
        if type(sockopts[0]) in (list, tuple):
            for sock_opt in sockopts[0]:
                level, option, value = sock_opt
                self.connection.sockopts.add((level, option, value))
        else:
            level, option, value = sockopts
            self.connection.sockopts.add((level, option, value))

    def _register_socket(self):
        """Create/Connect socket, apply options"""
        _socket = socket.socket(*Socket.protocols[self.connection.proto])
        _socket.settimeout(self.timeout)

        # Reconfigure original socket options.
        if self.connection.sockopts:
            for sock_opt in self.connection.sockopts:
                level, option, value = sock_opt
                _socket.setsockopt(level, option, value)

        _socket.connect(self.connection.connect_to)
        return _socket

    def _stream_send_and_receive(self, _socket, message, *args, **kwargs):
        """TCP/Stream sender and receiver"""
        data = BytesIO()

        _socket_fd = _socket.makefile(mode="rwb", encoding="utf-8")
        _socket_fd.write(message.encode("utf-8"))
        _socket_fd.flush()

        for received in _socket_fd:
            data.write(received)
        data.seek(0)

        _socket_fd.close()
        return data

    def _dgram_send_and_receive(self, _socket, message, buffer_size=4096, *args):
        """User Datagram Protocol sender and receiver"""
        _socket.send(message.encode("utf-8"))
        data, address = _socket.recvfrom(buffer_size)
        return BytesIO(data)

    def request(self, message, timeout=False, *args, **kwargs):
        """Populate connection pool, send message, return BytesIO, and cleanup"""
        if not self.connection_pool.full():
            self.connection_pool.put(self._register_socket())

        _socket = self.connection_pool.get()

        # setting timeout to None enables the socket to block.
        if timeout or timeout is None:
            _socket.settimeout(timeout)

        data = self.send_and_receive(_socket, message, *args, **kwargs)

        if self.connection.proto in Socket.streams:
            _socket.shutdown(socket.SHUT_RDWR)

        return Response(data, None, None)
